adage/analyze/management/commands/import_gene_network.py

```diff
@@ -153,31 +164,42 @@ def check_and_import(file_handle, ml_model):
         # If tokens[1] does not match one and only one gene's
         # systematic_name in the database, skip the line.
         try:
-            gene2 = find_gene(tokens[1])
+            gene2 = find_gene(tokens[1], cached_genes)
         except Exception as e:
             logger.warning("Input file line #%d will be skipped: %s",
                            line_index + 1, e)
             continue
 
         # Check whether the triplet (ml_model, gene1, gene2) is unique.
-        if not unique_together(ml_model, gene1, gene2):
+        if check_unique and not unique_together(ml_model, gene1, gene2):
             raise Exception("Input file line #%d: (%s, %s, %s) is not unique "
                             "in the database" % (line_index + 1,
                                                  tokens[0],
                                                  tokens[1],
                                                  ml_model.title))
 
-        # Import the valid data line into the database.
-        Edge.objects.create(mlmodel=ml_model, gene1=gene1, gene2=gene2,
-                            weight=weight)
+        # Save the valid data into batch.
+        records.append(
+            Edge(mlmodel=ml_model, gene1=gene1, gene2=gene2, weight=weight)
+        )
+        if len(records) == BULK_SIZE:  # dump bulk records into database
+            Edge.objects.bulk_create(records)
+            records = []
+    # Don't forget the last bulk:
+    if len(records) > 0:
+        Edge.objects.bulk_create(records)
+        records = []
 
 
-def find_gene(systematic_name):
+def find_gene(systematic_name, cached_genes):
     """
     Return the gene in database whose systematic_name matches input
     "systematic_name".  An exception will be raised if no gene is found
     or multiple genes exist in the database.
     """
+    if systematic_name in cached_genes:
+        return cached_genes[systematic_name]
```

Comment:
Interesting. So does this help speed things up as well? I would have thought that Django would do some QuerySet caching that would make the `Gene.objects.get` at line 204 work faster on subsequent calls with the same `systematic_name`. Maybe that's not how it works, though.

Reply:
Aha... okay, answered my own question. The `get()` method does not use a cache:

> https://docs.djangoproject.com/en/1.11/ref/models/querysets/#methods-that-do-not-return-querysets

so it would make sense that you would get a speedup from implementing your own cache.

Reply:
Yeah, the way I read the docs above, there is *no* caching on a Django `.objects.get()` call. This makes sense because you sometimes need to ensure you have the most up to date record.